### PR TITLE
Django 1.11 upgrade

### PR DESCRIPTION
**Pull request by davidfischer**
* Builds on #4319 but upgrades Django further to 1.11.16 (the latest LTS release)
* Uses the tastypie upgrade from #4325.
* Adds a data migration which sets [unusable passwords](https://docs.djangoproject.com/en/1.11/ref/contrib/auth/#django.contrib.auth.models.User.set_unusable_password) for anybody with old password schemes. They will need to request a password reset. These users haven't logged in since RTD upgraded to Django 1.4.
* Fixed a few minor test and lint failures due to changes in Django 1.11. There was a minor lint error due to the new [template based widget rendering](https://docs.djangoproject.com/en/2.0/releases/1.11/#changes-due-to-the-introduction-of-template-based-widget-rendering) and `Model.__init__` checking keyword values against known good values.

### CODE REVIEW

**Comment by davidfischer** [comment]
I should note that I have not yet extensively tested this. I ran the test suite, clicked around, and did a full local build but not much else.

**Comment by ericholscher** [comment]
Running this locally and will see if I hit any issues. Builds are working though, so that's good :)

File: readthedocs/rtd_tests/tests/test_api.py [hunk 1]
```diff
@@ -418,8 +418,8 @@ def test_get_invalid_raw_log(self):
 class APITests(TestCase):
     fixtures = ['eric.json', 'test_data.json']
 
-    def test_make_project(self):
-        """Test that a superuser can use the API."""
+    def test_cant_make_project(self):
+        """Test that a user can't use the API to create projects."""
```
Comment (humitos) [inline readthedocs/rtd_tests/tests/test_api.py]: You inverted the logic of the test here.

Don't we still need the test for testing that it's possible to use the API as super user to create a project?

File: readthedocs/settings/base.py [hunk 2]
```diff
@@ -126,7 +126,7 @@ def INSTALLED_APPS(self):  # noqa
     def USE_PROMOS(self):  # noqa
         return 'readthedocsext.donate' in self.INSTALLED_APPS
 
-    MIDDLEWARE_CLASSES = (
+    MIDDLEWARE = (
```
Comment (humitos) [inline readthedocs/settings/base.py]: This will require a change in `-ops` and `-corporate-ops`.

File: requirements/pip.txt [hunk 3]
```diff
@@ -16,12 +16,12 @@ mkdocs==0.17.3
...
 Markdown<3.0
 
-django==1.9.13
+django==1.11.16
 six==1.11.0
 future==0.16.0
 
-# django-tastypie 0.13.x and 0.14.0 are not compatible with our code
-django-tastypie==0.13.0
+# django-tastypie 0.14.0 drops support for django 1.9
```
Comment (humitos) [inline requirements/pip.txt]: If we are migrating to django 1.11 we may be possible to use 0.14.x here, I think.

**Comment by humitos** [comment]
I'm assuming that this PR closes #4174 #4319 #4325, right?

**Comment by humitos** [comment]
> at least I think we can test how much it breaks and start upgrading it there also.

It doesn't break _too much_. I've already opened a PR on the corporate site and all tests are passing. First step done.

File: readthedocs/rtd_tests/tests/test_api.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_api.py]: We decided to remove the creation of a project using the api, it was never documented anyway.

**Comment by stsewd** [comment]
> I'm assuming that this PR closes #4174 #4319 #4325, right?

Yup

File: readthedocs/rtd_tests/tests/test_api.py (hunk 1 above)
Comment (davidfischer) [reply readthedocs/rtd_tests/tests/test_api.py]: On this change specifically, there's more backstory in https://github.com/rtfd/readthedocs.org/pull/4325.

**Comment by ericholscher** [comment]
Just deployed 2.7.2. Will now merge this, so we can get more testing by using it locally & in CI.

**Pull request merged**